backend/src/main/kotlin/code/nebula/cipherquest/service/UserLevelService.kt

```diff
@@ -14,34 +17,74 @@ class UserLevelService(
+        private const val TIME_THRESHOLD = 30
+        private const val LEVEL_SCORE = 250
+        private const val MIN_ID = 1_000_000_000L
+        private const val MAX_ID = 9_999_999_999L
+    }
+
+    fun calculateScore(user: UserLevel): UserLevel {
+        val win = user.terminatedAt != null
+        val minutes = ChronoUnit.MINUTES.between(user.createdAt, user.terminatedAt ?: user.updatedAt)
+
+        val score =
+            (user.coins.times(COINS_SCORE))
+                .plus(if ((minutes - TIME_THRESHOLD) > 0) TIME_THRESHOLD - minutes else 0)
+                .plus((user.level.minus(1)).times(LEVEL_SCORE))
+                .plus(if (win) WIN_SCORE else 0)
+
+        user.score = score
+        return user
     }
 
     fun createUserLevel(request: CreateUserLevelRequest): UserLevel =
         userLevelRepository.save(
             UserLevel(
-                nextLong(1_000_000_000, 9_999_999_999).toString(),
+                userId = nextLong(MIN_ID, MAX_ID).toString(),
                 username = request.username,
                 level = DEFAULT_LEVEL,
             ),
         )
 
     fun getLevelByUser(userId: String): UserLevel =
-        userLevelRepository
-            .findById(userId)
-            .orElseThrow()
+        userLevelRepository.findById(userId).orElseThrow {
+            EntityNotFoundException("User with ID $userId not found")
+        }
 
     fun decreaseCoins(userId: String): UserLevel =
-        getLevelByUser(userId).apply { coins -= 1 }.let(userLevelRepository::save)
-
-    fun increaseLevelTo(userId: String, newLevel: Int): UserLevel =
-        getLevelByUser(userId).apply {
-            level = newLevel
-            coins += LEVEL_UP_COINS
-        }.let(userLevelRepository::save)
+        getLevelByUser(userId)
+            .apply { coins -= 1 }
+            .also(::calculateScore)
+            .let(userLevelRepository::save)
 
+    fun increaseLevelTo(
+        userId: String,
+        newLevel: Int,
+    ): UserLevel =
+        getLevelByUser(userId)
+            .apply {
+                level = newLevel
+                coins += LEVEL_UP_COINS
+            }.also(::calculateScore)
+            .let(userLevelRepository::save)
 
     fun hasWon(userId: String): UserLevel =
-        getLevelByUser(userId).apply {
-            terminatedAt = OffsetDateTime.now()
-        }.let(userLevelRepository::save)
+        getLevelByUser(userId)
+            .apply { terminatedAt = OffsetDateTime.now() }
+            .also(::calculateScore)
+            .let(userLevelRepository::save)
+
+    fun calculateScoreboard(): List<ScoreboardEntry> =
+        userLevelRepository
+            .findAll()
+            .map {
+                ScoreboardEntry(
+                    username = it.username,
+                    score = it.score.toInt(),
+                    userId = it.userId,
+                    time = ChronoUnit.MINUTES.between(it.createdAt, it.terminatedAt ?: it.updatedAt).toInt(),
```

Comment:
this is already used within the `calculateScore` method; maybe we could extract it in the userLevel class since it's not a real business logic or in an proper method inside this class